src/GraphQL/GraphQLBuilderExtensions.cs

```diff
@@ -0,0 +1,827 @@
+            builder.Register<TSchema>(serviceLifetime);
+
+            // Now register the service as ISchema if not already registered.
+            builder.TryRegister<ISchema, TSchema>(serviceLifetime);
+
+            return builder;
+        }
+
+        /// <summary>
+        /// Registers <paramref name="schema"/> within the dependency injection framework as <typeparamref name="TSchema"/>. <see cref="ISchema"/> is also
+        /// registered if it is not already registered within the dependency injection framework.
+        /// </summary>
+        public static IGraphQLBuilder AddSchema<TSchema>(this IGraphQLBuilder builder, TSchema schema)
+            where TSchema : class, ISchema
+            => schema == null ? throw new ArgumentNullException(nameof(schema)) : AddSchema(builder, _ => schema, ServiceLifetime.Singleton);
+
+        /// <inheritdoc cref="AddSchema{TSchema}(IGraphQLBuilder, ServiceLifetime)"/>
+        public static IGraphQLBuilder AddSchema<TSchema>(this IGraphQLBuilder builder, Func<IServiceProvider, TSchema> schemaFactory, ServiceLifetime serviceLifetime = ServiceLifetime.Singleton)
+            where TSchema : class, ISchema
+        {
+            if (schemaFactory == null)
+                throw new ArgumentNullException(nameof(schemaFactory));
+
+            if (serviceLifetime == ServiceLifetime.Transient && typeof(IDisposable).IsAssignableFrom(typeof(TSchema)))
+            {
+                // This scenario can cause a memory leak if the schema is requested from the root service provider.
+                // If it was requested from a scoped provider, then there is no reason to register it as transient.
+                // See following link:
+                // https://docs.microsoft.com/en-us/dotnet/core/extensions/dependency-injection-guidelines#disposable-transient-services-captured-by-container
+                throw new InvalidOperationException("A schema that implements IDisposable cannot be registered as a transient service.");
+            }
+
+            // Register the service with the DI provider as TSchema, overwriting any existing registration
+            builder.Register(schemaFactory, serviceLifetime);
+
+            // Now register the service as ISchema if not already registered.
+            builder.TryRegister<ISchema>(schemaFactory, serviceLifetime);
+
+            return builder;
+        }
+        #endregion
+
+        #region - AddDocumentExecuter -
+        /// <summary>
+        /// Registers <typeparamref name="TDocumentExecuter"/> as a singleton of type <see cref="IDocumentExecuter"/> within the
+        /// dependency injection framework.
+        /// </summary>
+        public static IGraphQLBuilder AddDocumentExecuter<TDocumentExecuter>(this IGraphQLBuilder builder)
+            where TDocumentExecuter : class, IDocumentExecuter
+            => builder.Register<IDocumentExecuter, TDocumentExecuter>(ServiceLifetime.Singleton);
```

Comment:
I see fixed lifetime. Considerations?

Reply:
Rather than allow builder configurations with every possible lifetime, I set up builder methods with all _recommended_ scenarios.  The GraphQL.NET document executer is designed to be a singleton.  If someone has a special need for a derived document executer that needs a different service lifetime, I would consider an advanced scenario and they can register it manually in their DI framework.  It's still only one line of code, and not even an additional line of code at that.

I did add support for registering schemas as scoped, because so many people still do that.

In summary:
- Schemas can be registered as singletons or scoped, and default to singletons
- Validation rules are registered as singletons.
- Middleware can be registered as a singleton or transient.  If transient, it will match the lifetime of the schema.  Transient is the default.
- Graph types are always registered as transients to match the service lifetime of the schema.  Since implementation, I have discovered that this can be an issue if specific instances of graph types are used in code, such as by `UnionGraphType.ResolveType`.  We could consider adding an overload to force a singleton registration, such as is allowed by AddMiddleware.
- DocumentCache are singletons
- DocumentWriter are singletons
- ErrorInfoProvider are singletons
- Complexity analyzer are singletons

Again, the idea is to encourage the recommend design pattern.  It is very easy to register any type with the underlying DI provider with a custom lifetime if there is a need for it.  The user will need to be aware of the consequences of selecting another lifetime, as some scenarios might cause problems.

It is also noteworthy that the builder is unnecessary for older code that manually registers the required services and graph types.  No new services are required for operation.  The entire builder system is a convenience, and does not hamper any existing code.

Reply:
OK